### PR TITLE
Update Exec and Portforward client to use pod subresource

### PR DESCRIPTION
**Pull request by csrwng**
Switches cli cmd for portfoward and exec to use /pods/[name]/exec and /pod/[name]/portforward

### CODE REVIEW

**Comment by csrwng** [comment]
@smarterclayton ptal

File: pkg/client/remotecommand/remotecommand.go [hunk 1]
```diff
@@ -66,26 +68,26 @@ func New(req *client.Request, config *client.Config, command []string, stdin io.
...
 // copied between these streams and the supplied stdin/stdout/stderr parameters.
-func (e *RemoteCommandExecutor) Execute() error {
+func (e *Executor) Execute() error {
 	doStdin := (e.stdin != nil)
 	doStdout := (e.stdout != nil)
 	doStderr := (!e.tty && e.stderr != nil)
 
 	if doStdin {
-		e.req.Param(api.ExecStdinParam, "1")
+		e.req.Param("stdin", "true")
```
Comment (smarterclayton) [inline pkg/client/remotecommand/remotecommand.go]: Where'd these constants go?  Deleted?  If so, you should be creating an api.ExecOptions and then converting it to the desired API version, then converting it to a map.

Comment (csrwng) [inline pkg/client/remotecommand/remotecommand.go]: @smarterclayton they weren't deleted, just not applicable anymore. They were for the internal endpoint. I will use the api.ExecOptions object to do the conversion.

**Comment by csrwng** [comment]
Updated parameter conversion for exec

**Comment by csrwng** [comment]
Moved conversion code under pkg/conversion

File: pkg/kubectl/cmd/portforward.go [hunk 2]
```diff
@@ -43,31 +44,58 @@ $ kubectl port-forward -p mypod 0:5000`
...
+	cmd.Flags().StringVarP(&pfw.pod, "pod", "p", "", "Pod name")
 	cmd.MarkFlagRequired("pod")
 	// TODO support UID
 	return cmd
 }
 
-func RunPortForward(f *cmdutil.Factory, cmd *cobra.Command, args []string) error {
-	podName := cmdutil.GetFlagString(cmd, "pod")
-	if len(podName) == 0 {
+type portForwardCmd struct {
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/portforward.go]: Why this abstraction?  Seems overly complex, not sure what you're trying to do here.

Comment (csrwng) [inline pkg/kubectl/cmd/portforward.go]: Only reason is to test... abstract out call to portforward.New(...) and following call to that object's ForwardPorts method. I'll think of a way to simplify it.

**Comment by csrwng** [comment]
complexity reduced

File: pkg/kubectl/cmd/portforward.go [hunk 3]
```diff
@@ -59,12 +60,25 @@ func NewCmdPortForward(f *cmdutil.Factory) *cobra.Command {
...
+
+func (*defaultPortForwarder) ForwardPorts(req *client.Request, config *client.Config, ports []string, stopChan <-chan struct{}) error {
+	fw, err := portforward.New(req, config, ports, stopChan)
+	if err != nil {
+		return err
+	}
+	return fw.ForwardPorts()
+}
+
+func runPortForward(f *cmdutil.Factory, cmd *cobra.Command, args []string, fw portForwarder) error {
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/portforward.go]: Why is this no longer public?

File: pkg/kubectl/cmd/exec.go [hunk 4]
```diff
@@ -39,36 +40,53 @@ $ kubectl exec -p 123456-7890 -c ruby-container -i -t -- bash -il`
...
+}
+
+type execParams struct {
+	podName       string
+	containerName string
+	stdin         bool
+	tty           bool
+}
+
+func runExec(f *cmdutil.Factory, cmd *cobra.Command, cmdIn io.Reader, cmdOut, cmdErr io.Writer, p *execParams, args []string, re remoteExecutor) error {
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/exec.go]: Should still be RunExec

File: pkg/kubectl/cmd/portforward.go (hunk 3 above)
Comment (csrwng) [inline pkg/kubectl/cmd/portforward.go]: doesn't need to be. It's only used by the command.

Comment (smarterclayton) [inline pkg/kubectl/cmd/portforward.go]: Our pattern is that all of these are public.

**Comment by csrwng** [comment]
reverted to public functions

**Comment by smarterclayton** [comment]
LGTM

**Pull request merged**